### PR TITLE
[CARBONDATA-2645] Segregate block and blocklet cache

### PR DESCRIPTION
**Pull request by manishgupta88**
Things done as part of this PR
1. Segregate block and blocklet cache. In this driver will cache the metadata based on CACHE_LEVEL. 
If CACHE_LEVEL is set to BLOCK then only carbondata files metadata will be cached in driver.
If CACHE_LEVEL is set to BLOCKLET thenmetadata  for  number of carbondata files * number of blocklets in each carbondata file will be cached in driver.

 - [ ] Any interfaces changed?
 No
 - [ ] Any backward compatibility impacted?
 No
 - [ ] Document update required?
Yes. will as handled in separate PR
 - [ ] Testing done
 Manually verified. UT will be added in another PR as part of jira  CARBONDATA-2649.
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6718/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5545/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5567/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5558/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6728/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5568/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5580/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6738/

File: core/src/main/java/org/apache/carbondata/core/indexstore/row/DataMapRow.java [hunk 1]
```diff
@@ -26,70 +26,66 @@
  */
 public abstract class DataMapRow implements Serializable {
 
-  protected CarbonRowSchema[] schemas;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/row/DataMapRow.java]: I don't think it is required to remove the schema from here as it is a temporary object to read data. It does not occupy any heap memory.  Many changes related to this is not required. Please revert the changes

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: use `import java.io.*`

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+      isLegacyStore = fileFooter.getBlockletList() == null;
+      // init segment properties and create schema
+      initSegmentProperties(fileFooter);
+      schemaBinary = convertSchemaToBinary(fileFooter.getColumnInTable());
+      createSchema(segmentProperties, blockletDataMapInfo.isAddToUnsafe(), true);
+      createSummarySchema(segmentProperties, schemaBinary, filePath, fileName, segmentId,
+          blockletDataMapInfo.isAddToUnsafe(), true);
+    }
+    // check for legacy store and load the metadata
+    DataMapRowImpl summaryRow = loadBlockMetadata(blockletDataMapInfo, indexInfo);
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: Create protected methods for `createSchema`, `createSummarySchema` and `loadMetadata` and override the same in BloclkletDataMap

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMap.java [hunk 4]
```diff
@@ -221,438 +159,71 @@ private DataMapRowImpl loadToUnsafe(DataFileFooter fileFooter,
...
-    }
-    return false;
-  }
-
   private List<Blocklet> prune(FilterResolverIntf filterExp, SegmentProperties segmentProperties) {
     if (memoryDMStore.getRowCount() == 0) {
       return new ArrayList<>();
     }
     List<Blocklet> blocklets = new ArrayList<>();
+    CarbonRowSchema[] schema = memoryDMStore.getSchema();
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMap.java]: Use the prune from super class

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapRowIndexes.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.indexstore.blockletindex;
+
+/**
+ * holder for blocklet info indexes in a DataMap row
+ */
+public class BlockletDataMapRowIndexes {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapRowIndexes.java]: Change to interface

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5608/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5598/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6776/

**Comment by manishgupta88** [comment]
@ravipesala ..handled review comments. Kindly review and merge

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6796/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5625/

**Comment by manishgupta88** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6846/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5642/

**Comment by ravipesala** [comment]
LGTM
